File: Mage/CoreData/Team.m
//
//  Team.m
//  mage-ios-sdk
//
//  Created by William Newman on 4/13/16.
//  Copyright © 2016 National Geospatial-Intelligence Agency. All rights reserved.
//

#import "Team.h"
#import "User.h"

@implementation Team

+ (Team *) insertTeamForJson: (NSDictionary *) json inManagedObjectContext:(NSManagedObjectContext *) context {
    Team *team = [Team MR_createEntityInContext:context];
    [team updateTeamForJson:json inManagedObjectContext:context];
    return team;
}

- (void) updateTeamForJson: (NSDictionary *) json inManagedObjectContext: (NSManagedObjectContext *) context{
    [self setRemoteId:[json objectForKey:@"id"]];
    [self setName:[json objectForKey:@"name"]];
    [self setTeamDescription:[json objectForKey:@"description"]];
    
    NSMutableSet *users = [[NSMutableSet alloc] init];
    for (NSString *userId in [json objectForKey:@"userIds"]) {
        // see if the user exists
        User *user = [User MR_findFirstByAttribute:@"remoteId" withValue:userId inContext:context];
        if (!user) {
            user = [User MR_createEntityInContext:context];
            [user setRemoteId:userId];
        }
        
        [users addObject:user];
    }
    
    [self setUsers:users];
}

@end

File: Mage/CoreData/Team.swift
//
//  Team.m
//  mage-ios-sdk
//
//  Created by William Newman on 4/13/16.
//  Copyright © 2016 National Geospatial-Intelligence Agency. All rights reserved.
//

import Foundation
import CoreData

@objc public class Team: NSManagedObject {
    
    @objc public static func insert(json: [AnyHashable : Any], context: NSManagedObjectContext) -> Team? {
        let team = Team.mr_createEntity(in: context);
        team?.update(json: json, context: context);
        return team;
    }
    
    @objc public func update(json: [AnyHashable : Any], context: NSManagedObjectContext) {
        self.remoteId = json[TeamKey.id.key] as? String
        self.name = json[TeamKey.name.key] as? String
        self.teamDescription = json[TeamKey.description.key] as? String
        
        var teamUsers: Set<User> = Set<User>()
        
        if let userIds = json[TeamKey.userIds.key] as? [String] {
            for userId in userIds {
                if let user = User.mr_findFirst(byAttribute: UserKey.remoteId.key, withValue: userId, in: context) {
                    teamUsers.insert(user);
                } else {
                    if let user = User.mr_createEntity(in: context) {
                        user.remoteId = userId;
                        teamUsers.insert(user);
                    }
                }
            }
        }
        
        self.users = teamUsers
    }
}
